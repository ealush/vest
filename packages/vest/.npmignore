src
playground
config
testUtils
./jest.config.js
dist