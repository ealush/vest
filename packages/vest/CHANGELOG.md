# vest - Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## 1.0.31 - 2021-07-02
### Fixed and improved
- e6ea7d6  support global replaces (undefined)
- 34e0414  improved conditions (undefined)
- 26c28c6  all tests pass (undefined)
- 33f4e46  release (undefined)
- 6fe40c7  better bundle (undefined)
- c2cfb65  better typing (undefined)
- c6387ab  before ts settings (undefined)
- c0e9708  generate correct d.ts file (undefined)
- packages/vest/types/vest.development.d.ts
- 8e01b8e  x (undefined)
- afb3960  x (undefined)
- e0a8463  add changelog support (undefined)
- cc46c38  current (undefined)
- 4a2073f  patch(vest): changes (ealush)
- c0a1705  patch(vest): changes (ealush)
- 240386c  better create generic (ealush)
- b6db1c6  transform any to unknowns (ealush)
- f2c20a7  add done integration tests (ealush)
- ba56e33  produce methods (ealush)
- 0638243  better produce integration tests (ealush)
- 4098a11  getFailures as integration tests (ealush)
- f56cf7f  pd (ealush)
- 2ea440e  pd (ealush)
- b3032ef  relocate draftresult (ealush)
- 962141f  test getFailuresByGroup (ealush)
- e163c56  test getFailures (ealush)
- a92b6df  test collectFailureMessages (ealush)
- 81aad51  fix most tests (ealush)
- c149e90  simplify collectFailureMessages (ealush)
