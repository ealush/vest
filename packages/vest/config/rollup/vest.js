const addEsmDir = require('../../../../config/rollup/addEsmDir');
const genConfig = require('../../../../config/rollup/genConfig');
const writeMainTemplate = require('../../../../config/rollup/writeMainTemplate');
const { packageDist, packageNames } = require('../../../../util');

addEsmDir(packageDist(packageNames.VEST));
writeMainTemplate(packageDist(packageNames.VEST), packageNames.VEST);

export default genConfig({
  distPath: packageDist(packageNames.VEST),
  esInput: 'vest.mjs.js',
  input: 'vest.js',
  libraryName: packageNames.VEST,
});
