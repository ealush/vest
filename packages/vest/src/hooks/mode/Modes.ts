export enum Modes {
  ALL,
  EAGER,
}
