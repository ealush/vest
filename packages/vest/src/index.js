import any from "anyone/any";
import enforce from "n4s/dist/enforce.min";
import { VERSION } from "./constants";
import test from "./core/test";
import validate from "./core/validate";
import { draft, only, skip, warn } from "./hooks";
import { singleton, runWithContext } from "./lib";

export default singleton.register({
  Enforce: enforce.Enforce,
  VERSION,
  any,
  draft,
  enforce,
  only,
  runWithContext,
  skip,
  test,
  validate,
  warn,
});
