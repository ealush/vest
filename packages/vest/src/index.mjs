import enforce from '../../n4s/src/enforce';
import create from './core/suite/create';
import test from './core/test';
import { draft, only, skip, warn, group } from './hooks';

const VERSION = VEST_VERSION;

export default {
  VERSION,
  create,
  draft,
  enforce,
  group,
  only,
  skip,
  test,
  warn,
};

export { VERSION, create, draft, enforce, group, only, skip, test, warn };
