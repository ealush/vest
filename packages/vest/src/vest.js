import create from 'createSuite';
import enforce from 'enforce';
import { only, skip, warn, group } from 'hooks';
import test from 'test';

const VERSION = __LIB_VERSION__;

export default {
  VERSION,
  create,
  enforce,
  group,
  only,
  skip,
  test,
  warn,
};
