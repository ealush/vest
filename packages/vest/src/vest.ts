import { enforce } from 'n4s';

import create from 'create';
import context from 'ctx';
import each from 'each';
import { only, skip } from 'exclusive';
import group from 'group';
import include from 'include';
import { eager } from 'mode';
import omitWhen from 'omitWhen';
import optional from 'optionalTests';
import skipWhen from 'skipWhen';
import { test } from 'test';
import warn from 'warn';

const VERSION = __LIB_VERSION__;
export {
  test,
  create,
  each,
  only,
  skip,
  warn,
  group,
  optional,
  skipWhen,
  omitWhen,
  enforce,
  VERSION,
  context,
  include,
  eager,
};
