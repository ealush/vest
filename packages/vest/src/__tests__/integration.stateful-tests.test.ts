import { enforce } from 'n4s';

import { Modes } from 'Modes';
import * as vest from 'vest';

describe('Stateful behavior', () => {
  let result;
  const validate = genSuite();

  test('Should merge skipped fields with previous values', () => {
    result = validate({ only: 'field_1' });
    expect(result.tests.field_1.errorCount).toBe(1);
    expect(result.errorCount).toBe(1);
    expect(Object.keys(result.tests)).toHaveLength(5); // including 4 skipped tests
    expect(result.tests).toHaveProperty('field_1');
    expect(result).toMatchSnapshot();

    result = validate({ only: 'field_5' });
    expect(result.errorCount).toBe(3);
    expect(result.tests.field_1.errorCount).toBe(1);
    expect(result.tests.field_5.errorCount).toBe(2);
    expect(Object.keys(result.tests)).toHaveLength(5); // including 4 skipped tests
    expect(result.tests).toHaveProperty('field_1');
    expect(result.tests).toHaveProperty('field_5');
    expect(result).toMatchSnapshot();

    result = validate();
    expect(result.errorCount).toBe(4);
    expect(result.tests.field_1.errorCount).toBe(1);
    expect(result.tests.field_2.errorCount).toBe(1);
    expect(result.tests.field_4.warnCount).toBe(1);
    expect(result.tests.field_5.errorCount).toBe(2);
    expect(Object.keys(result.tests)).toHaveLength(5);
    expect(result).toMatchSnapshot();
  });
});

describe('more complex', () => {
  const data: Record<string, string> = {};

  it('Should run correctly', () => {
    expect(suite.get().hasErrors()).toBe(false);
    expect(suite.get()).toMatchSnapshot();

    data.username = 'user_1';
    suite(data, 'username');

    expect(suite.get().hasErrors()).toBe(false);
    expect(suite.get()).toMatchSnapshot();

    suite(data, 'password');
    expect(suite.get().hasErrors()).toBe(true);
    expect(suite.get().tests.password).toMatchInlineSnapshot(`
      SummaryBase {
        "errorCount": 1,
        "errors": [
          "password is required",
        ],
        "pendingCount": 0,
        "testCount": 1,
        "valid": false,
        "warnCount": 0,
        "warnings": [],
      }
    `);
    expect(suite.get()).toMatchSnapshot();

    suite(data, 'confirm');
    expect(suite.get().tests.confirm).toMatchInlineSnapshot(`
      SummaryBase {
        "errorCount": 0,
        "errors": [],
        "pendingCount": 0,
        "testCount": 0,
        "valid": false,
        "warnCount": 0,
        "warnings": [],
      }
    `);
    expect(suite.get()).toMatchSnapshot();
    expect(suite.get().hasErrors('password')).toBe(true);
    expect(suite.get().hasErrors('confirm')).toBe(false);

    data.password = '123456';
    suite(data, 'password');
    expect(suite.get().tests.confirm).toMatchInlineSnapshot(`
      SummaryBase {
        "errorCount": 0,
        "errors": [],
        "pendingCount": 0,
        "testCount": 0,
        "valid": false,
        "warnCount": 0,
        "warnings": [],
      }
    `);
    data.confirm = '123456';
    suite(data, 'confirm');
    expect(suite.get().hasErrors('password')).toBe(false);
    expect(suite.get().hasErrors('confirm')).toBe(false);
    expect(suite.get().tests.confirm).toMatchInlineSnapshot(`
      SummaryBase {
        "errorCount": 0,
        "errors": [],
        "pendingCount": 0,
        "testCount": 1,
        "valid": true,
        "warnCount": 0,
        "warnings": [],
      }
    `);
  });

  const suite = vest.create(
    (data: Record<string, unknown> = {}, only: string) => {
      vest.only(only);

      vest.test('username', 'username is required', () => {
        enforce(data.username).isNotEmpty();
      });

      vest.test('username', 'username must be at least 3 characters', () => {
        enforce(data.username).longerThanOrEquals(3);
      });

      vest.test('password', 'password is required', () => {
        enforce(data.password).isNotEmpty();
      });

      vest.skipWhen(
        draft => draft.hasErrors('password'),
        () => {
          vest.test('confirm', 'passwords do not match', () => {
            enforce(data.confirm).equals(data.password);
          });
        }
      );
    }
  );
});

function genSuite() {
  return vest.create(({ only }: { only?: string | string[] } = {}) => {
    vest.mode(Modes.ALL);
    vest.only(only);
    vest.test('field_1', 'field_statement_1', () => false);
    vest.test('field_2', 'field_statement_2', () => {
      enforce(2).equals(3);
    });
    vest.test('field_3', 'field_statement_3', jest.fn());
    vest.test('field_4', 'field_statement_4', () => {
      vest.warn();
      throw new Error();
    });
    vest.test('field_4', 'field_statement_4', () => {
      vest.warn();
    });
    vest.test('field_5', 'field_statement_5', () => false);
    vest.test('field_5', 'field_statement_6', () => false);
  });
}
