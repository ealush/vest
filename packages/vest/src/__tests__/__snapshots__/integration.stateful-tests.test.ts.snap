// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stateful behavior Should merge skipped fields with previous values 1`] = `
Object {
  "done": [Function],
  "errorCount": 1,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "suiteName": undefined,
  "testCount": 1,
  "tests": Object {
    "field_1": Object {
      "errorCount": 1,
      "errors": Array [
        "field_statement_1",
      ],
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_2": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_3": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_4": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_5": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
  },
  "valid": false,
  "warnCount": 0,
}
`;

exports[`Stateful behavior Should merge skipped fields with previous values 2`] = `
Object {
  "done": [Function],
  "errorCount": 3,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "suiteName": undefined,
  "testCount": 3,
  "tests": Object {
    "field_1": Object {
      "errorCount": 1,
      "errors": Array [
        "field_statement_1",
      ],
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_2": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_3": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_4": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_5": Object {
      "errorCount": 2,
      "errors": Array [
        "field_statement_5",
        "field_statement_6",
      ],
      "testCount": 2,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
  },
  "valid": false,
  "warnCount": 0,
}
`;

exports[`Stateful behavior Should merge skipped fields with previous values 3`] = `
Object {
  "done": [Function],
  "errorCount": 4,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "suiteName": undefined,
  "testCount": 7,
  "tests": Object {
    "field_1": Object {
      "errorCount": 1,
      "errors": Array [
        "field_statement_1",
      ],
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_2": Object {
      "errorCount": 1,
      "errors": Array [
        "field_statement_2",
      ],
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_3": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 1,
      "valid": true,
      "warnCount": 0,
      "warnings": Array [],
    },
    "field_4": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 2,
      "valid": true,
      "warnCount": 1,
      "warnings": Array [
        "field_statement_4",
      ],
    },
    "field_5": Object {
      "errorCount": 2,
      "errors": Array [
        "field_statement_5",
        "field_statement_6",
      ],
      "testCount": 2,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
  },
  "valid": false,
  "warnCount": 1,
}
`;

exports[`more complex Should run correctly 1`] = `
Object {
  "errorCount": 0,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "suiteName": undefined,
  "testCount": 0,
  "tests": Object {},
  "valid": false,
  "warnCount": 0,
}
`;

exports[`more complex Should run correctly 2`] = `
Object {
  "errorCount": 0,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "suiteName": undefined,
  "testCount": 2,
  "tests": Object {
    "confirm": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "password": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "username": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 2,
      "valid": true,
      "warnCount": 0,
      "warnings": Array [],
    },
  },
  "valid": false,
  "warnCount": 0,
}
`;

exports[`more complex Should run correctly 4`] = `
Object {
  "errorCount": 1,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "suiteName": undefined,
  "testCount": 3,
  "tests": Object {
    "confirm": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "password": Object {
      "errorCount": 1,
      "errors": Array [
        "password is required",
      ],
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "username": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 2,
      "valid": true,
      "warnCount": 0,
      "warnings": Array [],
    },
  },
  "valid": false,
  "warnCount": 0,
}
`;

exports[`more complex Should run correctly 6`] = `
Object {
  "errorCount": 1,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "isValid": [Function],
  "isValidByGroup": [Function],
  "suiteName": undefined,
  "testCount": 3,
  "tests": Object {
    "confirm": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 0,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "password": Object {
      "errorCount": 1,
      "errors": Array [
        "password is required",
      ],
      "testCount": 1,
      "valid": false,
      "warnCount": 0,
      "warnings": Array [],
    },
    "username": Object {
      "errorCount": 0,
      "errors": Array [],
      "testCount": 2,
      "valid": true,
      "warnCount": 0,
      "warnings": Array [],
    },
  },
  "valid": false,
  "warnCount": 0,
}
`;
